### PR TITLE
Continue the Queue Mode implementation. Add client awake time feature.

### PR DESCRIPTION
**Pull request by carlosgp143**
This is the first of a set of pull request to introduce new features in the Queue Mode implementation, using #265 as starting point.

This first pull request makes two different things: make small changes to solve some problems that #265 has, and introduce the first feature: change the client state to SLEEPING after some time (`MAX_TRANSMIT_WAIT`).

**Main changes to 265**
1. First, I've removed the Observation Listener. Now the state of the client only depends on Registration and Update.
2. Perform all the implementation of the Queue Mode inside the `LwM2mQueue` class. In this class is where the new features can be added, for example the timer in this case. There is one object of this kind per client. As a result of this, there are some necessary changes inside the `PresenceService`.
3. Included in the `LeshanServer' the Queue Mode operation regarding request sending. If the client is sleeping, the 'send' methods return immediately and inform the user.

**New feature**
The feature added in this pull request is to set the client state to SLEEPING and notify the listeners when the time that the client states awake expires, which is `MAX_TRANSMIT_WAIT' as default. For it, a timer is started every time a new AWAKE event is triggered or a request is sent when the client is AWAKE, and when this timer expires, the state is changed.

Please note: the client state should change to SLEEPING also when a request is not responded by the client, but this feature will come with the new sender Pull Request.

### CODE REVIEW

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/LwM2mClientStateListener.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import org.eclipse.leshan.server.registration.RegistrationListener;
+import org.eclipse.leshan.server.registration.RegistrationUpdate;
+
+/**
+ * Listener that controls the state of the client (awake/sleeping) It is in charge of sending all the queued messages
+ * when the client is awake (has sent an update message), and controlling the time the client is awake before going to
+ * sleep.
+ */
+
+public class LwM2mClientStateListener implements RegistrationListener {
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/LwM2mClientStateListener.java]: Naming : I would prefer something like PresenceStateListener

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/LwM2mClientStateListener.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    LwM2mServer server;
+    PresenceService presenceService;
+
+    public LwM2mClientStateListener(LwM2mServer server) {
+        this.server = server;
+        presenceService = server.getPresenceService();
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/LwM2mClientStateListener.java]: It seems we don't need LwM2mServer as we only use `PresenceService`.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/LwM2mQueue.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+import org.eclipse.leshan.server.registration.Registration;
+import org.eclipse.leshan.util.Validate;
+
+/**
+ * Class that contains all the necessary elements to handle the queue mode. Every registration object that uses Queue
+ * Mode has a LwM2mQueue object for handling it.
+ */
+
+public class LwM2mQueue {
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/LwM2mQueue.java]: Naming : I would prefer something like `PresenceState` or `PresenceStatus.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/LwM2mQueue.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+    /* Elements to handle the time the client waits before going to sleep */
+    private Timer clientAwakeTimer;
+
+    private TimerTask clientAwakeTask;
+
+    private int clientAwakeTime;
+
+    /* Registration and Presence Service for notifying sleeping */
+    private Registration registration;
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/LwM2mQueue.java]: Registration should not be part of the state because object could change on update.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/LwM2mQueue.java [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        }
+
+    }
+
+    /**
+     * Reads from the Californium.properties file the MAX_TRANSMIT_WAIT constant of the CoAP Protocol
+     * 
+     * @return MAX_TRANSMIT_WAIT
+     */
+    public int readCoapMaxTransmitWait() {
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/LwM2mQueue.java]: Californium.properties is californium stuff and `leshan-server-core` does not know Californium.
This introduce an implicit dependency. 

We should replace this by a parameter or a Constant.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/LwM2mQueue.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    private Timer clientAwakeTimer;
+
+    private TimerTask clientAwakeTask;
+
+    private int clientAwakeTime;
+
+    /* Registration and Presence Service for notifying sleeping */
+    private Registration registration;
+
+    private PresenceService presenceService;
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/LwM2mQueue.java]: We should not have a `PresenceService` attribute. This create a cyclic dependency which is often means bad design.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/Presence.java [hunk 7]
```diff
@@ -20,7 +20,7 @@
  */
 public enum Presence {
     /** LWM2M Client is reachable and messages can be sent **/
-    ONLINE,
+    AWAKE,
     /** LWM2M Client is not reachable and no messages can be sent **/
-    OFFLINE
+    SLEEPING
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/Presence.java]: Naming: I would prefer ONLINE/OFFLINE but as the spec use both words, I think both are acceptable.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceService.java [hunk 8]
```diff
@@ -39,12 +40,34 @@
...
+     * @return true if the status is {@link Presence#SLEEPING}
+     */
+    boolean isClientSleeping(Registration registration);
+
+    /**
+     * Set the state of the client identified by registration as {@link Presence#AWAKE}
+     * 
+     * @param registration the client's registration object
+     */
+    public void setAwake(Registration registration);
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceService.java]: I think this is internal API and we don't want to expose it.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceService.java [hunk 9]
```diff
@@ -39,12 +40,34 @@
...
+
+    /**
+     * Notify the listeners that the client state changed to {@link Presence#SLEEPING}. The state changes is produced
+     * inside {@link LwM2mQueue} when the timer expires or when the client doesn't respond to a request.
+     * 
+     * @param registration the client's registration object
+     */
+    public void notifySleeping(Registration registration);
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceService.java]: I think this is internal API and we don't want to expose it.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceService.java [hunk 10]
```diff
@@ -39,12 +40,34 @@
...
+
+    /**
+     * Creates a new {@link LwM2mQueue} object associated with the client.
+     * 
+     * @param registration the client's registration object.
      */
-    boolean isOnline(Registration registration);
+    public void createQueueObject(Registration registration);
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceService.java]: I think this is internal API and we don't want to expose it.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java [hunk 11]
```diff
@@ -47,54 +45,43 @@ public void removeListener(PresenceListener listener) {
...
-     * 
-     * @param registration lwm2m client registration data
-     */
-    public void setOffline(Registration registration) {
-        if (registration.usesQueueMode()
-                && transitState(registration.getEndpoint(), Presence.ONLINE, Presence.OFFLINE)) {
+    @Override
+    public void setAwake(Registration registration) {
+        if (registration.usesQueueMode()) {
+            clientStatusList.get(registration.getEndpoint()).setAwake();
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java]: queue could be null.

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java [hunk 12]
```diff
@@ -400,6 +400,10 @@ public static void createAndStartServer(int webPort, String localAddress, int lo
             ServiceInfo coapSecureServiceInfo = ServiceInfo.create("_coaps._udp.local.", "leshan", secureLocalPort, "");
             jmdns.registerService(coapSecureServiceInfo);
         }
+        
+        /* Example of a listener that is notified every time the state of the client changes */
+        lwServer.getPresenceService().addListener(new ExamplePresenceListener("IPSO-objects-Firefly"));
+        lwServer.getPresenceService().addListener(new ExamplePresenceListener("hello"));
```
Comment (sbernard31) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java]: I suppose this should not be part of the PR ?

**Comment by carlosgp143** [comment]
Thanks for your feedback @sbernard31 ! Don't worry about the naming, I understand it's important.

I can do a commit with the small changes you have commented first, and then try to think about how to move logic from `LwM2mQueue` to `PresenceService` to have a better implementation. It's totally ok for me if you also refactor the code and come with a different PR. 

> I suspect we need to better handle multi-threading.

What do you mean with this? I can also take a look to that if you explain it a bit deeper...

**Comment by sbernard31** [comment]
> I can do a commit with the small changes you have commented first

Ok

>  then try to think about how to move logic from LwM2mQueue to PresenceService to have a better implementation

No problem, you can experiment.

> It's totally ok for me if you also refactor the code and come with a different PR.

Good to know :) (Sometime writing code is simpler than writing comments)

> What do you mean with this? I can also take a look to that if you explain it a bit deeper..

I feel that "state changing"+"timer task" should be done in a more atomic way but I did not think enough about that to give you a clear explanation.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java (hunk 11 above)
Comment (carlosgp143) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java]: How? When registration is created, the `LwM2mClientStateListener` creates a queue object inside the `PresenceServiceImpl` linked to the endpoint name, so it shouldn't be null... 

If you don't agree with that what would you do? Surround with try, catch or check first if it's not null and then call `setAwake()`?

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceService.java (hunk 10 above)
Comment (carlosgp143) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceService.java]: Do you mean these methods shouldn't be public, or that they shouldn't be part of the interface at all?

**Comment by carlosgp143** [comment]
Hi @sbernard31 !

I just made a new commit trying to move the logic from `PresenceStatus` (old `LwM2mQueue`) to `PresenceServiceImpl`.

The idea now is that `PresenceStatus` contains the state of the client, and the timer object, but not the methods to manage the timer. These methods have been moved to `PresenceServiceImpl`. My idea is to maintain this `PresenceStatus` object to put more Queue Mode features in the future, like for example the actual queues.

So now the `PresenceStateListener`, on every updated message, calls the `setAwake()` method of `PresenceServiceImpl` with the corresponding updated registration object as parameter. This object is used by `PresenceServiceImpl` to get the corresponding `PresenceStatus`object, change the state to AWAKE, get the timer and start it. I think this is also better for what you mentioned in:

> I feel that "state changing"+"timer task" should be done in a more atomic way but I did not think enough about that to give you a clear explanation." 

What do you think about this implementation?

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java [hunk 13]
```diff
@@ -47,54 +47,119 @@ public void removeListener(PresenceListener listener) {
     }
 
     @Override
-    public boolean isOnline(Registration registration) {
-        return Presence.ONLINE.equals(clientStatus.get(registration.getEndpoint()));
+    public boolean isClientSleeping(Registration registration) {
+        PresenceStatus presenceStatus = clientStatusList.get(registration.getEndpoint());
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java]: This is a public API, so user could use it with a Registration which does not exist anymore, right?
So how does we handle this ?

Personally,  I will go for `isPresent` / `isAwake` / `isOnline` instead of `isClientSleeping`.
and `isPresent` means we have a status for this registration and the status is present.
(About naming I sill feel, that we talk more about presence than "awakeness")

I even ask myself if this could be better to implement sleeping state by "no entry in the map" (but this is maybe too soon to go for this solution)

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java [hunk 14]
```diff
@@ -47,54 +47,119 @@ public void removeListener(PresenceListener listener) {
...
+        clientStatusList.put(reg.getEndpoint(), new PresenceStatus(clientAwakeTime));
+    }
+
+    /**
+     * Returns the {@link PresenceStatus} object associated with the given endpoint name.
+     * 
+     * @param reg The client's registration object.
+     * @return The {@link PresenceStatus} object.
+     */
+    public PresenceStatus getPresenceStatusObject(Registration reg) {
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java]: I think this should not be public, right ?

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java [hunk 15]
```diff
@@ -47,54 +47,119 @@ public void removeListener(PresenceListener listener) {
...
+    public void createPresenceStatusObject(Registration reg) {
+        clientStatusList.put(reg.getEndpoint(), new PresenceStatus());
+    }
+
+    /**
+     * Creates a new {@link PresenceStatus} object associated with the client, with a specific awake time.
+     * 
+     * @param registration the client's registration object.
+     */
+    public void createPresenceStatusObject(Registration reg, int clientAwakeTime) {
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java]: I ask myself, if this is not better to create automatically the status when it does not exist on setAwake ?

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/LeshanServer.java [hunk 16]
```diff
@@ -321,24 +282,57 @@ public LwM2mModelProvider getModelProvider() {
     @Override
     public <T extends LwM2mResponse> T send(Registration destination, DownlinkRequest<T> request)
             throws InterruptedException {
+        if (destination.usesQueueMode()) {
+            if (presenceService.isClientSleeping(destination)) {
+                LOG.info("The destination client is sleeping, request couldn't been sent.");
+                return null;
+            } else {
+                presenceService.startClientAwakeTimer(destination);
```
[the remark this answers is not in the document]
Comment (carlosgp143) [reply leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/LeshanServer.java]: A far as I have understood the LWM2M protocol, the client stays awake for `MAX_TRANXMIT_WAIT` after the last received request. So it will restart its internal timer on the reception of a request. So if we refresh the timer when we send the request, it will have a small drift with respect to the client's timer, due to the request transmit time. But if we restart it when we receive a response, it will have also a small drift due to the response transmit time, and the request processing time. So I think the question is what drift do we prefer? 

For me, it would be better to refresh it before sending the request, so that the server timer expires first and considers that the client is sleeping a little bit before it really is, rather than consider that it is sleeping a little bit after it really is.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java (hunk 13 above)
Comment (carlosgp143) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java]: I am using the Q-Mode with actual hardware IoT nodes, so for me it is more intuitive to use the AWAKE/SLEEPING naming. But if you really want it, I can change it to the ONLINE/OFFLINE one.

Apart from that I agree about "isPresent means we have a status for this registration and the status is present".

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/LeshanServer.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (sbernard31) [reply leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/LeshanServer.java]: But this is more complicated than that because this is not just : 
```
Request ->  
<- Response  
```
In CoAP, there is request acknowledgement which could be separated of response and even worst you could imagine scenario with blockwise transfer that means 1 LWM2M request means several CoAP requests.
([How timeout is handled](https://github.com/eclipse/leshan/wiki/Request-Timeout) give an idea of the complexity)

In the spec, nothing say that the client should response in MAX_TRANSMIT_WAIT.
It must wait MAX_TRANSMIT_WAIT after the last request before to go to sleep but between a request and a response time could be longer.

[the remark this answers is not in the document]
Comment (carlosgp143) [reply leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/LeshanServer.java]: Ok I understand it, I've simplified it too much. Then we agree that the client will restart its timer when it gets a request.

But then when should the server restart this timer? I don't see any good solution because:
1. If restart when the coap request is ACK, wouldn't work for non-confirmable messages.
2. If restart when response, as you said the time between the request and the response could be much more larger than the actual awake time of the client.
So... any other idea?

**Comment by carlosgp143** [comment]
New commit with some changes. You can check it! 

> Entry in PresenceServiceImpl map is never removed.

Now it is removed when the client is unregistered by the `PresenceStateListener`.

> We should probably enhance the threadsafeness of PresenceService but modification you have done will help a lot.

IF you give me some more tips about this I can tr to work on it...

> Not sure but Timer is a bit old school class maybe we should consider to use ScheduledThreadPoolExecutor instead.

It's true, I checked and they recommend to use ScheduledThreadPoolExecutor instead of Timer. So I changed the code in order to do that, it wasn't a big thing.

So I'm guessing that the only issues left are:
1. When to restart the timer?
2. What do we need to change to enhance the threadsafeness?

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/LeshanServer.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (sbernard31) [reply leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/LeshanServer.java]: FMPOV, we will have difficulty to exactly know when the device is here or not..

So we should keep in mind this will just be **an approximation**.
For me this approximation must be good enough to allow users to sends requests to the client and not so more.

So, 2 solutions : 
1. We set the client awake on Response. Setting awake means change state + relaunch timer.
=> This means, in some cases we could consider the client sleeping although it is awake, but once we receive the response we will consider it awake again and so could send the next requests.
2. We pause the timer before to send the request and reactive it when we receive the response.
=> So we will not go in sleeping mode waiting we receive the response but this could be long.

Personally, I will go the 1. approximation as this is the simplest one and we will see later if this approximation cause troubles.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceService.java [hunk 17]
```diff
@@ -47,5 +47,5 @@
      * @param registration data to check
      * @return true if the status is {@link Presence#SLEEPING}
      */
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceService.java]: javadoc is not OK anymore

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceStatus.java [hunk 18]
```diff
@@ -28,7 +30,9 @@
     private Presence state;
 
     /* Elements to handle the time the client waits before going to sleep */
-    private Timer clientAwakeTimer;
+    ScheduledExecutorService clientAwakeExecutor = Executors.newSingleThreadScheduledExecutor();
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceStatus.java]: Create 1 scheduler by device sounds so good.
This is more about creating 1 scheduler for all tasks in PresenceService.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceStatus.java [hunk 19]
```diff
@@ -70,8 +74,8 @@ public void setSleeping() {
      * 
      * @return true if client is sleeping
      */
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceStatus.java]: java doc

**Comment by sbernard31** [comment]
About Threadsafeness :

This is about being sure that we can not go in incoherent state, mainly because of race condition.

Often, the idea is to think about what should be done in an atomic way.
Then trying to use java concurrent tools to achieve this.

First, trying with simple tool like AtomicReference, ConcurentHashMap ...
If this is not enough, going with more complex code using  synchronized (Lock), ReadWriteLock ...

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceStatus.java (hunk 18 above)
Comment (carlosgp143) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceStatus.java]: But then you need to specify the number of tasks when creating the sheduler, and we don't now how many clients will be registered at a time... that's why I took this approach.

Comment (sbernard31) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceStatus.java]: This is not the number of tasks, this is more the number of workers.
A scheduler have 1 or severals workers/thread.
You can schedule more tasks than workers available.
The only risk is that task could be delay a bit.

Comment (carlosgp143) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceStatus.java]: Ok, I've misunderstood it. So then with just one single thread executor we can handle all the timer tasks of the registered clients, right?

Comment (sbernard31) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceStatus.java]: Yep

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/LeshanServer.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (carlosgp143) [reply leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/LeshanServer.java]: I totally agree that this will be an approximation. I think the first option is also better because it feels more safe for me to consider that the client is sleeping when it is actually awake than the other way around.

So I guess this implementation needs to go in the next pull request where the Queue Mode sender is added so that we can add the timer restarting inside the onResponse of the message observer. So I think I'll just remove the restarting of the timer from here and we will discuss more about it in that PR, once this one is accepted. Do you agree with that?

[the remark this answers is not in the document]
Comment (sbernard31) [reply leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/LeshanServer.java]: It sounds good to me :)

**Comment by carlosgp143** [comment]
I just made one commit changing to use just one Scheduled executor and to not restart the timer when a request is sent.

So I guess the only thing left in this PR is Threadsafeness. I only see that there can be problems when setting the client's state to awake or sleeping, and starting or stopping the timer. So... should we surround this actions with a Lock to ensure the threadsafeness? Otherwise I don't see anything else to do.

**Comment by sbernard31** [comment]
It sounds good.

> I only see that there can be problems when setting the client's state to awake or sleeping, and starting or stopping the timer. So... should we surround this actions with a Lock to ensure the threadsafeness? Otherwise I don't see anything else to do.

I agree. You can make it on this PR or we can also do that at the end (because the current design should allow us to do that later easily)

**Comment by carlosgp143** [comment]
Hi! I did it in this PR cause I think it's easier. 

So I guess this PR is ready to be merged into he queue_mode branch so that I move forward to the Queue Mode request sender... right? :-)

**Comment by sbernard31** [comment]
I will do that tomorrow.
You can start to go with a queue mode request sender. (Maybe an idea to reuse code is to create a sender which use the Current sender implementation using delegation)

**Comment by sbernard31** [comment]
Integrated in queue_mode branch.

**Comment by carlosgp143** [comment]
Hi! Just a couple of questions, cause I'm a bit confused...

1. Why didn't you integrate the last commit also? The "Added lock for atomic setAwake+starTimer and setSleeping+stopTimer" that I did yesterday?
2. Why did you close this PR and not merge it into the queue_mode branch? Maybe I'm a bit unfamiliar with the usual PR flow... but I don't know the difference.

**Comment by sbernard31** [comment]
1. because I just forgot it. Thx for you vigilance ! It is integrated now.
2. Just because I am used to integrate PR manually. :)

**Comment by carlosgp143** [comment]
Hey!

One question that comes to my mind when I think in how to implement the new Queue Mode Request Sender for the next PR. You said:

> You can start to go with a queue mode request sender. (Maybe an idea to reuse code is to create a sender which use the Current sender implementation using delegation)

Which, if I've understood correctly, means that the new sender will have the CaliforniumRequestSender as an attribute and will use its methods for sending requests. But inside this methods the CaliforniumRequestSender adds the corresponding Message observers, which are the ones we are interested in changing, for restarting the timer.

So... I don't think we can reuse much code... unless we change a bit the code of CaliforniumRequestSender and introduce the message observers as a parameter for the send methods. Otherwise I think the new sender needs to copy paste some code and add its own observers, in addition to check the state of the client to send or not. 

What's your opinion about this?

**Comment by sbernard31** [comment]
> Which, if I've understood correctly, means that the new sender will have the CaliforniumRequestSender as an attribute and will use its methods for sending requests. But inside this methods the CaliforniumRequestSender adds the corresponding Message observers, which are the ones we are interested in changing, for restarting the timer.

That's my idea.

>So... I don't think we can reuse much code... unless we change a bit the code of CaliforniumRequestSender and introduce the message observers as a parameter for the send methods. Otherwise I think the new sender needs to copy paste some code and add its own observers, in addition to check the state of the client to send or not.

I'm not sure to understand ?

What I have in mind should look like this : 
```java
 public <T extends LwM2mResponse> void send(final Registration destination, DownlinkRequest<T> request,
            final ResponseCallback<T> responseCallback, ErrorCallback errorCallback) {
       // Here do you specific test about state.
      // if you can send, use internal sender
        internalRequestSender.send(destination, request, new ResponseCallback<T>() {
            @Override
            public void onResponse(T response) {
                 // restart timer
                 // call the user callback.
                responseCallback.onResponse(response);
            }
        }, errorCallback);
     }
     // you probably need a custom error callback to support timeout and set the device sleeping.
     }
```

**Comment by carlosgp143** [comment]
Ok! I was thinking in using the methods with the SyncRequestObserver, but you're right, using the asynchronous ones makes it much easier.

So there should be a new PR quite soon :-)
